**로지스틱 회귀 분석**
- 사례) 합격 여부(Y:0, 1)을 공부시간(X1)과 수면시간(X2)으로 설명




In [ ]:
# 로지스틱 회귀분석: 종속변수가 범주형(이진)일 때, 독립변수가 종속변수 발생 확률에 미치는 영향 분석
# 모형식 출력: 확률(0~1사이의 값)

# 주요 지표
# 오즈비(odds ratio): 독립변수가 1 증가할 때, 사건 발생 odds가 몇 배 증가하는지
# 로그우도(log-likelihood): 모형 적합도 척도 ... 잔차이탈도 = -2*로그우도(작을수록 모형 적합이 좋음)

import pandas as pd

df = pd.DataFrame({
    "공부시간": [5, 6, 7, 8, 2, 1, 9, 10, 4, 3,
              6, 8, 7, 9, 2, 3, 5, 10, 8, 4,
              7, 6, 9, 2, 5, 8, 4, 6, 7, 3],
    "수면시간": [6, 7, 8, 5, 9, 10, 4, 3, 8, 9,
              7, 5, 6, 4, 10, 9, 8, 3, 5, 8,
              6, 7, 4, 10, 8, 5, 9, 7, 6, 9],
    "합격여부": [1, 1, 1, 0, 0, 0, 1, 1, 0, 0,
              1, 1, 0, 1, 0, 0, 1, 1, 1, 0,
              1, 0, 1, 0, 1, 1, 0, 1, 0, 0]
})

1. 로지스틱 회귀분석 수행

In [ ]:
import statsmodels.formula.api as smf

# 선형 회귀랑 ols -> logit만 다름
model = smf.logit("합격여부 ~ 공부시간 + 수면시간", data=df).fit()

print(model.summary())

# 모형적합도: Pseudo R-squared...두 변수로 합격 여부 변동을 38.6% 설명
# 공부시간 0.9475: 공부시간이 1증가할 때 로그 우즈가 약 0.95 증가
# 수면시간
# 둘 다 유의하지 않음

# 결론: 공부/수면 시간이 합격여부에 영향이 없진 않으나 통계적으로 유의하지 않음

Optimization terminated successfully.
         Current function value: 0.424105
         Iterations 6
                           Logit Regression Results                           
Dep. Variable:                   합격여부   No. Observations:                   30
Model:                          Logit   Df Residuals:                       27
Method:                           MLE   Df Model:                            2
Date:                Tue, 09 Jun 2026   Pseudo R-squ.:                  0.3862
Time:                        18:34:29   Log-Likelihood:                -12.723
converged:                       True   LL-Null:                       -20.728
Covariance Type:            nonrobust   LLR p-value:                 0.0003339
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -6.6344      9.450     -0.702      0.483     -25.156      11.888
공부시간           0.9475      0.

2. 결과해석

(1) 모형 적합도
 - Pseudo R-squared = 0.3862
  → 공부시간과 수면시간 두 변수를 통해 합격 여부의 변동을 약 38.6%
설명 가능

(2) 회귀계수 (coef)
 - Intercept = -6.6344 (p=0.483, 유의하지 않음)
 - 공부시간 = 0.9475 (p=0.167, 유의하지 않음)
  → 공부시간이 1 증가할 때 합격할 log odds가 약 0.95 증가
 - 수면시간 = 0.1903 (p=0.815, 유의하지 않음)
 → 수면시간이 1 증가할 때 합격할 log odds가 약 0.19 증가

(3) 결론
 -  공부시간과 수면시간이 합격여부에 영향을 주는 경향은 있으나,
표본 수 및 데이터 특성상 통계적으로 유의하지 않음


3. 오즈비 계산

In [3]:
import numpy as np
# 넘파이의 exponential
odds_ratios = np.exp(model.params)
print("오즈비:\n", odds_ratios)

오즈비:
 Intercept    0.001314
공부시간         2.579259
수면시간         1.209573
dtype: float64


- 공부시간 : 2.579
  → 공부시간이 1 증가하면 합격할 오즈가 약 2.58배 증가
- 수면시간 : 1.210
  → 수면시간이 1 증가하면 합격할 오즈가 약 1.21배 증가


4. 잔차이탈도 계산

In [ ]:
llf = model.llf  # 로그우도 log likelihood
deviance = -2 * llf # 잔차이탈도는 로그우드 *(-2)
print("로그우도:", llf)
print("잔차이탈도:", deviance)

로그우도: -12.723152944196704
잔차이탈도: 25.446305888393407
